*** configure2	Fri Oct  7 21:20:59 2011
--- configure2	Sat Apr 28 13:54:54 2012
***************
*** 103,118 ****
  test -z "$shlibdir" && shlibdir=$PREFIX/lib
  test -z "$incdir" && incdir=$PREFIX/include/dvdnav
  
- dvdread=no
- $DVDREAD_CONFIG --prefix >> /dev/null 2>&1 && dvdread=yes
- test "$dvdread" != "yes" && $DVDREAD_CONFIG --exists >> /dev/null 2>&1 && dvdread=yes
- if test "$dvdread" != "yes" ; then
-   echo "$DVDREAD_CONFIG returned an error. Can't proceed"
-   exit 1
- fi
- dvdread_cflags=`$DVDREAD_CONFIG --cflags`
- dvdread_libs=`$DVDREAD_CONFIG --libs`
- 
  targetos=`uname -s`
  case $targetos in
    Darwin)
--- 103,108 ----
