--- upnp/src/genlib/miniserver/miniserver.c	2013-11-15 16:50:02.000000000 +0100
+++ upnp/src/genlib/miniserver/miniserver.c	2014-01-07 14:50:05.674342546 +0100
@@ -549,6 +549,9 @@
 		return UPNP_E_SOCKET_BIND;
 	}
 #endif
+/* LXiMedia: Do not override with ephemeral ports as this will create conflicts
+ *  every now and then with dynamically assigned ports. */
+#if 0
 	/* As per the IANA specifications for the use of ports by applications
 	 * override the listen port passed in with the first available. */
 	if (listen_port4 < APPLICATION_LISTENING_PORT) {
@@ -559,6 +562,7 @@
 		listen_port6 = (uint16_t)APPLICATION_LISTENING_PORT;
 	}
 #endif
+#endif /* End LXiMedia */
 	memset(&__ss_v4, 0, sizeof (__ss_v4));
 	serverAddr4->sin_family = (sa_family_t)AF_INET;
 	serverAddr4->sin_addr.s_addr = htonl(INADDR_ANY);
